filemanager/s3manager_v2.go

```diff
@@ -0,0 +1,326 @@
+	"github.com/rudderlabs/rudder-go-kit/logger"
+)
+
+const ServiceName = "s3"
+
+type S3ManagerV2 struct {
+	*baseManager
+	config *S3Config
+
+	sessionConfig *awsutil.SessionConfig
+	client        *s3.Client
+	configMu      sync.Mutex
+}
+
+// NewS3Manager creates a new file manager for S3
+func NewS3ManagerV2(
+	config map[string]interface{}, log logger.Logger, defaultTimeout func() time.Duration,
+) (*S3ManagerV2, error) {
+	var s3Config S3Config
+	if err := mapstructure.Decode(config, &s3Config); err != nil {
+		return nil, err
+	}
+
+	sessionConfig, err := awsutil.NewSimpleSessionConfig(config, ServiceName)
+	if err != nil {
+		return nil, err
+	}
+
+	s3Config.RegionHint = appConfig.GetString("AWS_S3_REGION_HINT", "us-east-1")
+
+	return &S3ManagerV2{
+		baseManager: &baseManager{
+			logger:         log,
+			defaultTimeout: defaultTimeout,
+		},
+		config:        &s3Config,
+		sessionConfig: sessionConfig,
+	}, nil
+}
+
+func (m *S3ManagerV2) ListFilesWithPrefix(ctx context.Context, startAfter, prefix string, maxItems int64) ListSession {
+	return &s3ListSessionV2{
+		baseListSession: &baseListSession{
+			ctx:        ctx,
+			startAfter: startAfter,
+			prefix:     prefix,
+			maxItems:   maxItems,
+		},
+		manager:     m,
+		isTruncated: true,
+	}
+}
+
+// Download downloads a file from S3
+func (m *S3ManagerV2) Download(ctx context.Context, output *os.File, key string) error {
+	client, err := m.getClient(ctx)
+	if err != nil {
+		return fmt.Errorf("s3 client: %w", err)
+	}
+
+	downloader := s3manager.NewDownloader(client)
+
+	ctx, cancel := context.WithTimeout(ctx, m.getTimeout())
+	defer cancel()
+
+	_, err = downloader.Download(ctx, output,
+		&s3.GetObjectInput{
+			Bucket: aws.String(m.config.Bucket),
+			Key:    aws.String(key),
+		})
+
+	if err != nil {
+		var nsk *types.NoSuchKey
+		if errors.As(err, &nsk) {
+			// handle NoSuchKey error
+			return ErrKeyNotFound
+		}
+		return err
+	}
+	return nil
+}
+
+// Upload uploads a file to S3
+func (m *S3ManagerV2) Upload(ctx context.Context, file *os.File, prefixes ...string) (UploadedFile, error) {
+	fileName := path.Join(m.config.Prefix, path.Join(prefixes...), path.Base(file.Name()))
+
+	uploadInput := &s3.PutObjectInput{
+		ACL:    types.ObjectCannedACLBucketOwnerFullControl,
+		Bucket: aws.String(m.config.Bucket),
+		Key:    aws.String(fileName),
+		Body:   file,
+	}
+
+	if m.config.EnableSSE {
+		uploadInput.ServerSideEncryption = types.ServerSideEncryptionAes256
+	}
+
+	client, err := m.getClient(ctx)
+	if err != nil {
+		return UploadedFile{}, fmt.Errorf("s3 client: %w", err)
+	}
+	uploader := s3manager.NewUploader(client)
+
+	ctx, cancel := context.WithTimeout(ctx, m.getTimeout())
+	defer cancel()
+
+	output, err := uploader.Upload(ctx, uploadInput)
+	if err != nil {
+		if codeErr, ok := err.(codeError); ok && codeErr.Code() == "MissingRegion" {
+			err = fmt.Errorf(fmt.Sprintf(`Bucket '%s' not found.`, m.config.Bucket))
+		}
```

Comment:
Is it possible to use `aws.MissingRegionError`?
```suggestion
                var regionError *aws.MissingRegionError
		if errors.As(err, &regionError) {
			err = fmt.Errorf(fmt.Sprintf(`Bucket '%s' not found.`, m.config.Bucket))
		}
```